### ISSUE TITLE
Type errors from combining Expressions with numbers

### ISSUE THREAD

**Issue by carlosgmartin** [cvxpy/cvxpy#2242]
**Describe the bug**

Combining expressions of type [`cp.Expression`](https://www.cvxpy.org/api_reference/cvxpy.expressions.html) with ones of type `int` and `float` raises type errors in [pyright](https://github.com/microsoft/pyright).

**To Reproduce**

Let `test.py` contain
```python3
import cvxpy as cp

x = cp.Variable()
_ = x + 1
_ = x - 1
_ = x * 1
_ = x / 1
_ = x <= 1
_ = x >= 1
_ = x == 1
```
Now run
```
$ pyright test.py
```

**Expected behavior**

```
0 errors, 0 warnings, 0 informations
```

**Output**

```
test.py
  test.py:3:8 - error: "Variable" is not exported from module "cvxpy" (reportPrivateImportUsage)
  test.py:4:5 - error: Operator "+" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
  test.py:5:5 - error: Operator "-" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
  test.py:6:5 - error: Operator "*" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
  test.py:7:5 - error: Operator "/" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
  test.py:9:5 - error: Operator "<=" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
  test.py:10:5 - error: Operator ">=" not supported for types "Variable" and "Literal[1]" (reportGeneralTypeIssues)
7 errors, 0 warnings, 0 informations
```

For some reason, the `==` case doesn't raise a type error.

**Version**
 - OS: macOS 11.7.10
 - CVXPY Version: 1.3.2
 - Pyright Version: 1.1.329

**Additional context**

These errors can be suppressed by using [`cp.Constant`](https://www.cvxpy.org/api_reference/cvxpy.expressions.html#cvxpy.expressions.constants.Constant), as in `cp.Constant(1)`. However, the code runs correctly as-is, and adding `cp.Constant`s everywhere in complex expressions is cumbersome.

There is also another error:
```
/Users/carlos/Desktop/test.py
  /Users/carlos/Desktop/test.py:3:8 - error: "Variable" is not exported from module "cvxpy" (reportPrivateImportUsage)
```
This can be fixed by exporting the names of [`cvxpy/cvxpy/__init__.py`](https://github.com/cvxpy/cvxpy/blob/e8c2cc28f59d7346cdd14f56728d837955df4c5d/cvxpy/__init__.py). Example:
```python3
from cvxpy.expressions.variable import Variable as Variable
```

**Comment by rileyjmurray** [comment]
As you’ve clearly found out, CVXPY doesn’t support static type checking at
the moment. We’ve slowly been adding type annotations throughout the code
base. We’d welcome more contributions in this regard if you’d be interested
in helping!

**Comment by carlosgmartin** [comment]
@SteveDiamond I think this should stay open, since there's still the issue of the Expression type errors.

**Comment by carlosgmartin** [comment]
@rileyjmurray @SteveDiamond The type error disappears if I replace all occurrences of `other: "Expression"` with `other`—that is, drop the type annotation—for the `Expression` class methods in [`cvxpy.expressions.expression`](https://github.com/cvxpy/cvxpy/blob/master/cvxpy/expressions/expression.py). Another alternative might be to replace `other: "Expression"` with `other: ExpressionLike` where `ExpressionLike` is a union of `Expression` and any types that [`cast_op`](https://github.com/cvxpy/cvxpy/blob/master/cvxpy/expressions/expression.py#L46) can accept, such as `int`, `float`, numpy array, etc. What do you think?

**Comment by SteveDiamond** [comment]
> @rileyjmurray @SteveDiamond The type error disappears if I replace all occurrences of `other: "Expression"` with `other`—that is, drop the type annotation—for the `Expression` class methods in [`cvxpy.expressions.expression`](https://github.com/cvxpy/cvxpy/blob/master/cvxpy/expressions/expression.py). Another alternative might be to replace `other: "Expression"` with `other: ExpressionLike` where `ExpressionLike` is a union of `Expression` and any types that [`cast_op`](https://github.com/cvxpy/cvxpy/blob/master/cvxpy/expressions/expression.py#L46) can accept, such as `int`, `float`, numpy array, etc. What do you think?

ExpressionLike sounds better to me.

**Comment by rileyjmurray** [comment]
Seems like this has been resolved by #2250. Thanks for the contribution, @carlosgmartin!

(If anyone things I'm closing this issue prematurely, please speak up.)
